File: tessoku_book/a21.cpp
#include <iostream>
#include <vector>
#include<algorithm>
using namespace std;
typedef long long ll;
#define rep0(i,n) for(int i=0;i<=n;i++)
#define rep(i,n) for(int i=1;i<=n;i++)
int n,p[2009],a[2009],scorel,scorer;
int dp[2009][2009];

int main(){
    cin>>n;
    rep0(i,2008)rep0(j,2008)dp[i][j]=0;
    dp[1][n]=0;
    rep(i,n)cin>>p[i]>>a[i];
    for(int i=1;i<=n;i++)for(int j=n;j>=1;j--){
        scorer=0;
        scorel=0;
        if(i>=2 && i<=p[i-1]&& p[i-1]<=j )scorel=a[i-1];
        if(j<=n-1 && i<=p[j+1]&&p[j+1]<=j)scorer=a[j+1]; 
        if(i==1)dp[i][j]=dp[i][j+1]+scorer;
        else if(j==n)dp[i][j]=dp[i-1][j]+scorel;
        else dp[i][j]=max(dp[i-1][j]+scorel,dp[i][j+1]+scorer);
    }
    int answer=0;
    rep(i,n)answer=max(answer,dp[i][i]);
    cout<<answer<<endl;
    return 0;
}

File: tessoku_book/a21.py
n = int(input())
p = [0 for _ in range(n+1)]
a = [0 for _ in range(n+1)]

for i in range(n):
    p[i+1],a[i+1] = map(int,input().split())
    
dp = [[0 for _ in range(n+2)] for _ in range(n+2)]

dp[1][n] = 0

for r in reversed(range(1, n+1)):
    for l in range(1, n+1):
        scorel = 0
        if l >= 2 and l <= p[l-1] and p[l-1] <= r:
            scorel = a[l-1]
            
        scorer = 0
        if r <= n-1 and l <= p[r+1] and p[r+1] <= r:
            scorer = a[r+1]
        
        if l == 1:
            dp[l][r] = dp[l][r+1] + scorer
        elif r == n:
            dp[l][r] = dp[l-1][r] + scorel
        else:
            dp[l][r] = max(dp[l-1][r] + scorel, dp[l][r+1] + scorer)

answer = 0
for i in range(1,n+1):
    answer = max(answer, dp[i][i])

print(answer)